File: app/src/main/java/de/herrmann_engel/rbv/activities/FileTools.java
package de.herrmann_engel.rbv.activities;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.provider.DocumentsContract;
import android.view.View;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.documentfile.provider.DocumentFile;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import de.herrmann_engel.rbv.Globals;
import de.herrmann_engel.rbv.R;
import de.herrmann_engel.rbv.adapters.AdapterMediaLinkCardAll;
import de.herrmann_engel.rbv.adapters.AdapterMediaLinkCardImages;
import de.herrmann_engel.rbv.databinding.DiaConfirmBinding;
import de.herrmann_engel.rbv.databinding.DiaImageBinding;
import de.herrmann_engel.rbv.databinding.DiaRecBinding;
import de.herrmann_engel.rbv.db.DB_Media;
import de.herrmann_engel.rbv.db.DB_Media_Link_Card;
import de.herrmann_engel.rbv.db.utils.DB_Helper_Get;

public abstract class FileTools extends AppCompatActivity {


    final ActivityResultLauncher<Intent> selectMediaFolder = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    try {
                        Intent data = result.getData();
                        Uri destination = data.getData();
                        setCardMediaFolder(destination.toString());
                        int takeFlags = data.getFlags();
                        takeFlags &= (Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
                        getContentResolver().takePersistableUriPermission(destination, takeFlags);
                        notifyFolderSet();
                    } catch (Exception e) {
                        e.printStackTrace();
                        Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show();
                    }
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            String folder = getCardMediaFolder();
            DB_Helper_Get dbHelperGet = new DB_Helper_Get(this);
            //Check: No media folder set
            if (folder == null || folder.isEmpty()) {
                if (!dbHelperGet.getAllMedia().isEmpty()) {
                    showSelectDialog(getResources().getString(R.string.select_folder_help_db_entries));
                }
            } else {
                DocumentFile outputDirectory = DocumentFile.fromTreeUri(this, Uri.parse(folder));
                //Check: Media folder set but directory not accessible
                if (outputDirectory == null || !outputDirectory.isDirectory()) {
                    showSelectDialog(getResources().getString(R.string.select_folder_help_reselect));
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    protected void startSelectMediaFolder() {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        intent.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION);
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
        selectMediaFolder.launch(intent);
    }

    protected void showSelectDialog(String text) {
        Dialog selectFolderInfoDialog = new Dialog(this, R.style.dia_view);
        DiaConfirmBinding bindingSelectFolderInfoDialog = DiaConfirmBinding.inflate(getLayoutInflater());
        selectFolderInfoDialog.setContentView(bindingSelectFolderInfoDialog.getRoot());
        selectFolderInfoDialog.setTitle(getResources().getString(R.string.select_folder_title));
        selectFolderInfoDialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT);
        bindingSelectFolderInfoDialog.diaConfirmDesc.setText(text);
        bindingSelectFolderInfoDialog.diaConfirmDesc.setVisibility(View.VISIBLE);
        bindingSelectFolderInfoDialog.diaConfirmYes.setOnClickListener(v -> {
            startSelectMediaFolder();
            selectFolderInfoDialog.dismiss();
        });
        bindingSelectFolderInfoDialog.diaConfirmNo.setOnClickListener(v -> selectFolderInfoDialog.dismiss());
        selectFolderInfoDialog.show();
    }

    public void showMissingDialog(int card) {
        Dialog selectFolderInfoDialog = new Dialog(this, R.style.dia_view);
        DiaConfirmBinding bindingSelectFolderInfoDialog = DiaConfirmBinding.inflate(getLayoutInflater());
        selectFolderInfoDialog.setContentView(bindingSelectFolderInfoDialog.getRoot());
        selectFolderInfoDialog.setTitle(getResources().getString(R.string.media_missing_dialog));
        selectFolderInfoDialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT);
        bindingSelectFolderInfoDialog.diaConfirmDesc.setText(getResources().getString(R.string.media_missing_dialog_text));
        bindingSelectFolderInfoDialog.diaConfirmDesc.setVisibility(View.VISIBLE);
        bindingSelectFolderInfoDialog.diaConfirmYes.setOnClickListener(v -> {
            selectFolderInfoDialog.dismiss();
            notifyMissingAction(card);
        });
        bindingSelectFolderInfoDialog.diaConfirmNo.setOnClickListener(v -> selectFolderInfoDialog.dismiss());
        selectFolderInfoDialog.show();
    }

    public void showImageDialog(int id) {
        Dialog imageDialog = new Dialog(this, R.style.dia_view);
        DiaImageBinding bindingImageDialog = DiaImageBinding.inflate(getLayoutInflater());
        imageDialog.setContentView(bindingImageDialog.getRoot());
        imageDialog.setTitle(getResources().getString(R.string.image_media));
        imageDialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT);
        Picasso.get().load(getImageUri(id)).fit().centerInside().into(bindingImageDialog.diaImageView);
        imageDialog.show();
    }

    public void showImageListDialog(ArrayList<DB_Media_Link_Card> imageList) {
        if (imageList.size() == 1) {
            showImageDialog(imageList.get(0).file);
        } else if (imageList.size() > Globals.IMAGE_PREVIEW_MAX) {
            showMediaListDialog(imageList, true, getResources().getString(R.string.query_media_image_title));
        } else {
            Dialog imageListDialog = new Dialog(this, R.style.dia_view);
            DiaRecBinding bindingImageListDialog = DiaRecBinding.inflate(getLayoutInflater());
            imageListDialog.setContentView(bindingImageListDialog.getRoot());
            imageListDialog.setTitle(getResources().getString(R.string.query_media_image_title));
            imageListDialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                    WindowManager.LayoutParams.MATCH_PARENT);
            imageListDialog.show();
            AdapterMediaLinkCardImages adapter = new AdapterMediaLinkCardImages(imageList);
            bindingImageListDialog.diaRec.setAdapter(adapter);
            bindingImageListDialog.diaRec.setLayoutManager(new GridLayoutManager(this, 3));
        }
    }

    public void showMediaListDialog(ArrayList<DB_Media_Link_Card> mediaList) {
        showMediaListDialog(mediaList, false, getResources().getString(R.string.query_media_all_title));
    }

    private void showMediaListDialog(ArrayList<DB_Media_Link_Card> mediaList, boolean onlyImages, String title) {
        Dialog mediaListDialog = new Dialog(this, R.style.dia_view);
        DiaRecBinding bindingMediaListDialog = DiaRecBinding.inflate(getLayoutInflater());
        mediaListDialog.setContentView(bindingMediaListDialog.getRoot());
        mediaListDialog.setTitle(title);
        mediaListDialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT);
        mediaListDialog.show();
        AdapterMediaLinkCardAll adapter = new AdapterMediaLinkCardAll(mediaList, onlyImages);
        bindingMediaListDialog.diaRec.setAdapter(adapter);
        bindingMediaListDialog.diaRec.setLayoutManager(new LinearLayoutManager(this));
    }

    public Dialog showDeleteDialog(String fileName, String text) {
        Dialog deleteFileDialog = new Dialog(this, R.style.dia_view);
        DiaConfirmBinding bindingDeleteFileDialog = DiaConfirmBinding.inflate(getLayoutInflater());
        deleteFileDialog.setContentView(bindingDeleteFileDialog.getRoot());
        deleteFileDialog.setTitle(getResources().getString(R.string.delete_file_title));
        deleteFileDialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT);
        bindingDeleteFileDialog.diaConfirmDesc.setText(text);
        bindingDeleteFileDialog.diaConfirmDesc.setVisibility(View.VISIBLE);
        bindingDeleteFileDialog.diaConfirmYes.setOnClickListener(v -> {
            deleteMediaFile(fileName);
            deleteFileDialog.dismiss();
        });
        bindingDeleteFileDialog.diaConfirmNo.setOnClickListener(v -> deleteFileDialog.dismiss());
        deleteFileDialog.show();
        return deleteFileDialog;
    }

    public Dialog showDeleteDialog(String fileName) {
        return showDeleteDialog(fileName, getResources().getString(R.string.delete_file_info));
    }

    public void handleNoMediaFile() {
        try {
            String folder = getCardMediaFolder();
            if (folder == null || folder.isEmpty()) {
                return;
            }
            DocumentFile outputDirectory = DocumentFile.fromTreeUri(this, Uri.parse(folder));
            if (outputDirectory == null || !outputDirectory.exists() || !outputDirectory.isDirectory()) {
                return;
            }
            SharedPreferences settings = getSharedPreferences(Globals.SETTINGS_NAME, MODE_PRIVATE);
            boolean mediaInGallery = settings.getBoolean("media_in_gallery", true);
            DocumentFile noMediaFile = outputDirectory.findFile(".nomedia");
            if (mediaInGallery && noMediaFile != null && noMediaFile.exists() && noMediaFile.isFile()) {
                DocumentsContract.deleteDocument(
                        getContentResolver(),
                        noMediaFile.getUri()
                );
            } else if (!mediaInGallery && (noMediaFile == null || !noMediaFile.exists())) {
                outputDirectory.createFile("application/octet-stream", ".nomedia");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public Uri getImageUri(int id) {
        DocumentFile folderFile = DocumentFile.fromTreeUri(this, Uri.parse(getCardMediaFolder()));
        if (folderFile == null) {
            return null;
        }
        DocumentFile file = getFile(id);
        if (file == null) {
            return null;
        }
        return file.getUri();
    }

    public DocumentFile[] listFiles() {
        DocumentFile folderFile = DocumentFile.fromTreeUri(this, Uri.parse(getCardMediaFolder()));
        if (folderFile == null) {
            return null;
        }
        return folderFile.listFiles();
    }

    private DocumentFile getFile(String name) {
        DocumentFile folderFile = DocumentFile.fromTreeUri(this, Uri.parse(getCardMediaFolder()));
        if (folderFile == null) {
            return null;
        }
        return folderFile.findFile(name);
    }

    private DocumentFile getFile(int id) {
        DB_Helper_Get dbHelperGet = new DB_Helper_Get(this);
        DB_Media currentMedia = dbHelperGet.getSingleMedia(id);
        if (currentMedia == null) {
            return null;
        }
        String fileName = currentMedia.file;
        return getFile(fileName);
    }

    private void openFile(DocumentFile file) {
        Intent intent = new Intent();
        intent.setAction(android.content.Intent.ACTION_VIEW);
        intent.setDataAndType(file.getUri(), file.getType());
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        this.startActivity(intent);
    }

    public void openFile(String name) {
        try {
            DocumentFile file = getFile(name);
            openFile(file);
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show();
        }
    }

    public void openFile(int id) {
        try {
            DocumentFile file = getFile(id);
            openFile(file);
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show();
        }
    }

    private void shareFile(DocumentFile file) {
        Intent intent = new Intent();
        intent.setAction(Intent.ACTION_SEND);
        intent.setType(file.getType());
        intent.putExtra(Intent.EXTRA_STREAM, file.getUri());
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        this.startActivity(Intent.createChooser(intent, file.getName()));
    }

    public void shareFile(String name) {
        try {
            DocumentFile file = getFile(name);
            shareFile(file);
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show();
        }
    }

    public void shareFile(int id) {
        try {
            DocumentFile file = getFile(id);
            shareFile(file);
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show();
        }
    }

    private void deleteMediaFile(String name) {
        try {
            DocumentFile file = getFile(name);
            if (file != null && file.isFile()) {
                DocumentsContract.deleteDocument(
                        getContentResolver(),
                        file.getUri()
                );
            } else {
                Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show();
            }
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show();
        }
    }

    public String getCardMediaFolder() {
        SharedPreferences config = this.getSharedPreferences(Globals.CONFIG_NAME, Context.MODE_PRIVATE);
        return config.getString("card_media_folder", null);
    }

    private void setCardMediaFolder(String uri) {
        SharedPreferences config = this.getSharedPreferences(Globals.CONFIG_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = config.edit();
        editor.putString("card_media_folder", uri);
        editor.apply();
    }

    public boolean existsMediaFile(String name) {
        try {
            DocumentFile folderFile = DocumentFile.fromTreeUri(this, Uri.parse(getCardMediaFolder()));
            return folderFile.findFile(name).exists();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public boolean existsMediaFile(int input) {
        DB_Helper_Get dbHelperGet = new DB_Helper_Get(this);
        DB_Media currentMedia = dbHelperGet.getSingleMedia(input);
        if (currentMedia == null) {
            return false;
        }
        String fileName = currentMedia.file;
        return existsMediaFile(fileName);
    }

    protected abstract void notifyFolderSet();

    protected abstract void notifyMissingAction(int id);
}

File: app/src/main/java/de/herrmann_engel/rbv/activities/FileTools.kt
package de.herrmann_engel.rbv.activities

import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.DocumentsContract
import android.view.View
import android.view.WindowManager
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.edit
import androidx.documentfile.provider.DocumentFile
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.squareup.picasso.Picasso
import de.herrmann_engel.rbv.Globals
import de.herrmann_engel.rbv.R
import de.herrmann_engel.rbv.adapters.AdapterMediaLinkCardAll
import de.herrmann_engel.rbv.adapters.AdapterMediaLinkCardImages
import de.herrmann_engel.rbv.databinding.DiaConfirmBinding
import de.herrmann_engel.rbv.databinding.DiaImageBinding
import de.herrmann_engel.rbv.databinding.DiaRecBinding
import de.herrmann_engel.rbv.db.DB_Media
import de.herrmann_engel.rbv.db.utils.DB_Helper_Get

abstract class FileTools : AppCompatActivity() {
    private val selectMediaFolder = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == RESULT_OK) {
            val data = result.data
            val destination = data?.data
            var takeFlags = data?.flags
            if (takeFlags != null && destination != null) {
                setCardMediaFolder(destination.toString())
                takeFlags =
                    takeFlags and (Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
                contentResolver.takePersistableUriPermission(destination, takeFlags)
                notifyFolderSet()
            } else {
                Toast.makeText(this, R.string.error, Toast.LENGTH_LONG).show()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        try {
            val folder = cardMediaFolder
            val dbHelperGet = DB_Helper_Get(this)
            // Check: No media folder set
            if (folder.isNullOrEmpty()) {
                if (dbHelperGet.allMedia.isNotEmpty()) {
                    showSelectDialog(resources.getString(R.string.select_folder_help_db_entries))
                }
            } else {
                val outputDirectory = DocumentFile.fromTreeUri(this, Uri.parse(folder))
                // Check: Media folder is set, but directory not accessible
                if (outputDirectory == null || !outputDirectory.isDirectory) {
                    showSelectDialog(resources.getString(R.string.select_folder_help_reselect))
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun startSelectMediaFolder() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
        intent.addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intent.addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
        selectMediaFolder.launch(intent)
    }

    protected fun showSelectDialog(text: String?) {
        val selectFolderInfoDialog = Dialog(this, R.style.dia_view)
        val bindingSelectFolderInfoDialog = DiaConfirmBinding.inflate(
            layoutInflater
        )
        selectFolderInfoDialog.setContentView(bindingSelectFolderInfoDialog.root)
        selectFolderInfoDialog.setTitle(resources.getString(R.string.select_folder_title))
        selectFolderInfoDialog.window!!.setLayout(
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.MATCH_PARENT
        )
        bindingSelectFolderInfoDialog.diaConfirmDesc.text = text
        bindingSelectFolderInfoDialog.diaConfirmDesc.visibility = View.VISIBLE
        bindingSelectFolderInfoDialog.diaConfirmYes.setOnClickListener {
            startSelectMediaFolder()
            selectFolderInfoDialog.dismiss()
        }
        bindingSelectFolderInfoDialog.diaConfirmNo.setOnClickListener { selectFolderInfoDialog.dismiss() }
        selectFolderInfoDialog.show()
    }

    private fun showMissingDialog(card: Int, parentDialog: Dialog?) {
        val selectFolderInfoDialog = Dialog(this, R.style.dia_view)
        val bindingSelectFolderInfoDialog = DiaConfirmBinding.inflate(
            layoutInflater
        )
        selectFolderInfoDialog.setContentView(bindingSelectFolderInfoDialog.root)
        selectFolderInfoDialog.setTitle(resources.getString(R.string.media_missing_dialog))
        selectFolderInfoDialog.window!!.setLayout(
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.MATCH_PARENT
        )
        bindingSelectFolderInfoDialog.diaConfirmDesc.text =
            resources.getString(R.string.media_missing_dialog_text)
        bindingSelectFolderInfoDialog.diaConfirmDesc.visibility = View.VISIBLE
        bindingSelectFolderInfoDialog.diaConfirmYes.setOnClickListener {
            parentDialog?.dismiss()
            selectFolderInfoDialog.dismiss()
            notifyMissingAction(card)
        }
        bindingSelectFolderInfoDialog.diaConfirmNo.setOnClickListener { selectFolderInfoDialog.dismiss() }
        selectFolderInfoDialog.show()
    }

    fun showImageDialog(fileId: Int, cardId: Int, dialog: Dialog? = null) {
        val uri = getImageUri(fileId)
        if (uri != null) {
            val imageDialog = Dialog(this, R.style.dia_view)
            val bindingImageDialog = DiaImageBinding.inflate(
                layoutInflater
            )
            imageDialog.setContentView(bindingImageDialog.root)
            imageDialog.setTitle(resources.getString(R.string.image_media))
            imageDialog.window!!.setLayout(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT
            )
            Picasso.get().load(uri).fit().centerInside()
                .into(bindingImageDialog.diaImageView)
            imageDialog.show()
        } else if (!existsMediaFile(fileId)) {
            showMissingDialog(cardId, dialog)
        }
    }

    protected fun showImageListDialog(imageList: ArrayList<DB_Media>, cardNo: Int) {
        if (imageList.size == 1) {
            showImageDialog(imageList[0].uid, cardNo)
        } else if (imageList.size > Globals.MAX_SIZE_CARD_IMAGE_PREVIEW) {
            showMediaListDialog(
                imageList,
                cardNo,
                true,
                resources.getString(R.string.query_media_image_title)
            )
        } else {
            val imageListDialog = Dialog(this, R.style.dia_view)
            val bindingImageListDialog = DiaRecBinding.inflate(
                layoutInflater
            )
            imageListDialog.setContentView(bindingImageListDialog.root)
            imageListDialog.setTitle(resources.getString(R.string.query_media_image_title))
            imageListDialog.window!!.setLayout(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT
            )
            imageListDialog.show()
            val adapter = AdapterMediaLinkCardImages(imageList, cardNo, imageListDialog)
            bindingImageListDialog.diaRec.adapter = adapter
            bindingImageListDialog.diaRec.layoutManager = GridLayoutManager(this, 3)
        }
    }

    protected fun showMediaListDialog(mediaList: ArrayList<DB_Media>, cardNo: Int) {
        showMediaListDialog(mediaList, cardNo, false, resources.getString(R.string.query_media_all_title))
    }

    private fun showMediaListDialog(
        mediaList: ArrayList<DB_Media>,
        cardNo: Int,
        onlyImages: Boolean,
        title: String
    ) {
        val mediaListDialog = Dialog(this, R.style.dia_view)
        val bindingMediaListDialog = DiaRecBinding.inflate(
            layoutInflater
        )
        mediaListDialog.setContentView(bindingMediaListDialog.root)
        mediaListDialog.setTitle(title)
        mediaListDialog.window!!.setLayout(
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.MATCH_PARENT
        )
        mediaListDialog.show()
        val adapter = AdapterMediaLinkCardAll(mediaList, cardNo, onlyImages, mediaListDialog)
        bindingMediaListDialog.diaRec.adapter = adapter
        bindingMediaListDialog.diaRec.layoutManager = LinearLayoutManager(this)
    }

    @JvmOverloads
    fun showDeleteDialog(
        fileName: String,
        text: String? = resources.getString(R.string.delete_file_info)
    ): Dialog {
        val deleteFileDialog = Dialog(this, R.style.dia_view)
        val bindingDeleteFileDialog = DiaConfirmBinding.inflate(
            layoutInflater
        )
        deleteFileDialog.setContentView(bindingDeleteFileDialog.root)
        deleteFileDialog.setTitle(resources.getString(R.string.delete_file_title))
        deleteFileDialog.window!!.setLayout(
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.MATCH_PARENT
        )
        bindingDeleteFileDialog.diaConfirmDesc.text = text
        bindingDeleteFileDialog.diaConfirmDesc.visibility = View.VISIBLE
        bindingDeleteFileDialog.diaConfirmYes.setOnClickListener {
            deleteMediaFile(fileName)
            deleteFileDialog.dismiss()
        }
        bindingDeleteFileDialog.diaConfirmNo.setOnClickListener { deleteFileDialog.dismiss() }
        deleteFileDialog.show()
        return deleteFileDialog
    }

    protected fun handleNoMediaFile() {
        try {
            val folder = cardMediaFolder
            if (folder.isNullOrEmpty()) {
                return
            }
            val outputDirectory = DocumentFile.fromTreeUri(this, Uri.parse(folder))
            if (outputDirectory == null || !outputDirectory.exists() || !outputDirectory.isDirectory) {
                return
            }
            val settings = getSharedPreferences(Globals.SETTINGS_NAME, MODE_PRIVATE)
            val mediaInGallery = settings.getBoolean("media_in_gallery", true)
            val noMediaFile = outputDirectory.findFile(".nomedia")
            if (mediaInGallery && noMediaFile != null && noMediaFile.exists() && noMediaFile.isFile) {
                DocumentsContract.deleteDocument(
                    contentResolver,
                    noMediaFile.uri
                )
            } else if (!mediaInGallery && (noMediaFile == null || !noMediaFile.exists())) {
                outputDirectory.createFile("application/octet-stream", ".nomedia")
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getImageUri(id: Int): Uri? {
        DocumentFile.fromTreeUri(this, Uri.parse(cardMediaFolder)) ?: return null
        val file = getFile(id) ?: return null
        return file.uri
    }

    protected fun listFiles(): Array<DocumentFile>? {
        val folderFile = DocumentFile.fromTreeUri(this, Uri.parse(cardMediaFolder))
            ?: return null
        return folderFile.listFiles()
    }

    private fun getFile(name: String): DocumentFile? {
        val folderFile = DocumentFile.fromTreeUri(this, Uri.parse(cardMediaFolder))
            ?: return null
        return folderFile.findFile(name)
    }

    private fun getFile(id: Int): DocumentFile? {
        val dbHelperGet = DB_Helper_Get(this)
        val currentMedia = dbHelperGet.getSingleMedia(id) ?: return null
        val fileName = currentMedia.file
        return getFile(fileName)
    }

    private fun openFile(file: DocumentFile) {
        try {
            val intent = Intent()
            intent.action = Intent.ACTION_VIEW
            intent.setDataAndType(file.uri, file.type)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            this.startActivity(intent)
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show()
        }
    }

    fun openFile(name: String) {
        val file = getFile(name)
        if (file != null && file.isFile) {
            openFile(file)
        } else {
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show()
        }
    }

    fun openFile(id: Int) {
        val file = getFile(id)
        if (file != null && file.isFile) {
            openFile(file)
        } else {
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show()
        }
    }

    private fun shareFile(file: DocumentFile) {
        val intent = Intent()
        intent.action = Intent.ACTION_SEND
        intent.type = file.type
        intent.putExtra(Intent.EXTRA_STREAM, file.uri)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        this.startActivity(Intent.createChooser(intent, file.name))
    }

    fun shareFile(name: String) {
        val file = getFile(name)
        if (file != null && file.isFile) {
            shareFile(file)
        } else {
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show()
        }
    }

    fun shareFile(id: Int) {
        val file = getFile(id)
        if (file != null && file.isFile) {
            shareFile(file)
        } else {
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show()
        }
    }

    private fun deleteMediaFile(name: String) {
        val file = getFile(name)
        if (file != null && file.isFile) {
            DocumentsContract.deleteDocument(
                contentResolver,
                file.uri
            )
        } else {
            Toast.makeText(this, R.string.error, Toast.LENGTH_SHORT).show()
        }
    }

    protected val cardMediaFolder: String?
        get() {
            val config = getSharedPreferences(Globals.CONFIG_NAME, MODE_PRIVATE)
            return config.getString("card_media_folder", null)
        }

    private fun setCardMediaFolder(uri: String) {
        val config = getSharedPreferences(Globals.CONFIG_NAME, MODE_PRIVATE)
        config.edit {
            putString("card_media_folder", uri)
        }
    }

    fun existsMediaFile(name: String?): Boolean {
        val folderFile = DocumentFile.fromTreeUri(
            this, Uri.parse(
                cardMediaFolder
            )
        )
        return name?.let { folderFile?.findFile(it)?.exists() } == true
    }

    fun existsMediaFile(input: Int): Boolean {
        val dbHelperGet = DB_Helper_Get(this)
        val currentMedia = dbHelperGet.getSingleMedia(input) ?: return false
        val fileName = currentMedia.file
        return existsMediaFile(fileName)
    }

    protected abstract fun notifyFolderSet()
    protected abstract fun notifyMissingAction(id: Int)
}
